### PR TITLE
WSOTWOCLOUD-730- Fixing [Identity Cloud] Dynamically create the topli…

### PR DESCRIPTION
**Pull request by dimuthud**
…nk menu using a config file

### CODE REVIEW

File: jaggeryapps/admin/templates/header.jag [hunk 1]
```diff
@@ -33,15 +33,31 @@ if (userInfo) {
 }
 const USERPORTAL_LINK = "/user-portal/t/" + tenantDomain;
 var headerConfig;
+var identityMgtUrl;
+var mainMenu;
+var userMenu;
+var expandMenu;
+var tenantDisplayName;
+
+headerConfig  = readFile("/conf/toplink_menu.json");
```
Comment (malithie) [inline jaggeryapps/admin/templates/header.jag]: So every time header gets loaded we read the file. Can you check whether this is pattern followed in other cases also

File: jaggeryapps/admin/conf/toplink_menu.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
+{
+  "Identity":{
+    "url": "https://cloudmgt.cloud.wso2.com/cloudmgt/site/pages/index.jag"
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/conf/toplink_menu.json]: Do we need to add the wso2 production urls to this?
Ideally these should be applied from the puppet and these values should be kept in hieradata and changed according to the environment.

File: jaggeryapps/admin/templates/header.jag [hunk 3]
```diff
@@ -33,15 +33,31 @@ if (userInfo) {
...
+  mainMenu = headerConfig.Main;
+  userMenu = headerConfig.User;
+  expandMenu = headerConfig.Expand;
+
+  for (var attr in mainMenu) {
+     var dropDowns = mainMenu[attr].dropDown;
+     var icon = mainMenu[attr].icon;
+     var url = mainMenu[attr].url;
+     var target = mainMenu[attr].target;
+         if(attr == "Domain"){
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/templates/header.jag]: formatting issue.

File: jaggeryapps/admin/templates/header.jag [hunk 4]
```diff
@@ -73,67 +88,107 @@ function resolveUserName(user) {
...
-                <li><a href="<%=headerConfig.support%>"><i class="fa fa-envelope"></i> Support</a></li>
-                <li class="dropdown user-name">
-                    <a href="#" class="dropdown-toggle" data-toggle="dropdown"><%=encode.forHtml(resolveUserName(MultitenantUtils.getTenantAwareUsername(session.get(LOGGED_IN_USER))))%>
+                <%
+                    for (var attr in mainMenu) {
+                        var dropDowns = mainMenu[attr].dropDown;
+                        var icon = mainMenu[attr].icon;
+                        var url = mainMenu[attr].url;
+                        var target = mainMenu[attr].target;
+                        if(attr == "Domain"){
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/templates/header.jag]: formatting issue.

File: jaggeryapps/admin/templates/header.jag [hunk 5]
```diff
@@ -73,67 +88,107 @@ function resolveUserName(user) {
...
+                            attr = tenantDomain;
+                        }
+                %>
+                <li class="dropdown">
+                    <%
+                        if(dropDowns == "false"){
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/templates/header.jag]: formatting issue.

File: jaggeryapps/admin/templates/header.jag [hunk 6]
```diff
@@ -73,67 +88,107 @@ function resolveUserName(user) {
...
+                    %>
+                    <a href="<%=url%>"  target="<%=target%>"><i class="<%=icon%>"></i> <%=attr%></a>
+                    <%
+                        }else{
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/templates/header.jag]: formatting issue.

**Review by madurangasiriwardena — CHANGES_REQUESTED** [review]

File: jaggeryapps/admin/conf/toplink_menu.json (hunk 2 above)
Comment (dimuthud) [inline jaggeryapps/admin/conf/toplink_menu.json]: Url has been replace with "#"

File: jaggeryapps/admin/templates/header.jag (hunk 1 above)
Comment (dimuthud) [inline jaggeryapps/admin/templates/header.jag]: Introduced a new method getJsonFromFile(..).

File: jaggeryapps/admin/templates/header.jag (hunk 3 above)
Comment (dimuthud) [inline jaggeryapps/admin/templates/header.jag]: Fixed.

File: jaggeryapps/admin/templates/header.jag (hunk 5 above)
Comment (dimuthud) [inline jaggeryapps/admin/templates/header.jag]: Fixed

File: jaggeryapps/admin/templates/header.jag (hunk 4 above)
Comment (dimuthud) [inline jaggeryapps/admin/templates/header.jag]: Fixed

File: jaggeryapps/admin/templates/header.jag (hunk 6 above)
Comment (dimuthud) [inline jaggeryapps/admin/templates/header.jag]: Fixed.

**Review by madurangasiriwardena — APPROVED** [review]

File: jaggeryapps/admin/templates/header.jag [hunk 7]
```diff
@@ -53,13 +69,36 @@ function resolveUserName(user) {
...
+
+        try{
+            var file = new File(pathTofile);
+            file.open("r");
+            var data = file.readAll();
+            cloudConfig = parse(data);
+            file.close();
+            application.put(pathTofile,cloudConfig);
+        } catch (e) {
+            log.error("Error while reading file " + pathTofile + "\n" + e);
```
Comment (madurangasiriwardena) [inline jaggeryapps/admin/templates/header.jag]: Can you check the exception handling logic here? Do we need to log and throw? If it is printed like this, will it print the stacktrace?

**Pull request merged**